### PR TITLE
QueryLibrary model and reducer

### LINKED ISSUE #1198: Query library data model
Design a data model which can accommodate the query library constraints we intend to implement initially, while also being flexible enough to adapt to some of the future features we have discussed.

**Initial requirements**
- [ ] model can easily serialize into (or simply is) `json` so it may be easily transported/shared.
- [ ] metadata
  * tags: many to many cardinality for tags :: queries
  * name: possibly limited in length
  * description
  * value: actual zql to be executed
- [ ] parametric search (description, name, value), though perhaps just by name initially? (cc @alfred-landrum)
- [ ] tag filtering
- [ ] ordered (and reorder-able) while not filtered

**Future features**
- [ ] query templates, or parametric queries: allow for parameters which are meant to be filled in by the user in addition to selecting the query itself
- [ ] grouping: directory style, arbitrary, user-defined/named grouping of queries
- [ ] versioning: for synchronizing with remote sources

Finally, this ticket should also include writing the corresponding reducer for the model that we come up with

### PR DESCRIPTION
**Pull request by mason-fish**
fixes #1198 

Adds a new QueryLibrary reducer. This code is all additive and is not yet being used, though it will be once we link together the remaining ui work tracked in #1199 

The underlying data structures we are using to represent this lib are simply an object tree and arrays. The arrays provide the order which we need while also keep the structure easily serializable into JSON, and the tree structure provides the grouping hierarchy which we will want to use later on. The issue/inefficiency of re copying elements on array resize/reorder/remove is moot since we are using redux, which depends on data being immutable and when changes occur they must be done by creating new resources anyway.

Signed-off-by: Mason Fish <mason@looky.cloud>

### CODE REVIEW

File: src/js/state/QueryLibrary/selectors.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import {QueryLibraryState} from "./types"
+import {State} from "../types"
+
+export default {
+  getRaw: (state: State): QueryLibraryState => state.queryLibrary
+}
```
Comment (mason-fish) [inline src/js/state/QueryLibrary/selectors.ts:6]: We may want more selectors here, and will likely return the library as a parsed tree model in one of them. For now I am only returning the raw state and the tests are all built off of that, and as I am wiring this up to the components that consume this data then I may add more

File: src/js/state/QueryLibrary/initial.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import {QueryLibraryState} from "./types"
+
+const init = (): QueryLibraryState => ({
+  id: "root",
+  name: "root",
+  items: []
```
Comment (mason-fish) [inline src/js/state/QueryLibrary/initial.ts:6]: this will actually be populated by #1201

File: src/js/state/QueryLibrary/reducer.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  const srcItem = get(draft, toItemPath(srcItemPath), null)
+
+  if (!srcItem) return
+  if (!get(draft, toItemPath(initial(destItemPath)), null)) return
+
+  // If the move is all in the same directory then the adjusting indices can
+  // cause an off by one issue since the destination index will be affected after
+  // removal (e.g. an item cannot be moved to the end of its current group because of this).
+  // For this situation we instead remove the item first, and then insert its copy
+  if (srcItemPath.length === destItemPath.length) {
```
Comment (jameskerr) [inline src/js/state/QueryLibrary/reducer.ts:74]: Is this just comparing the depths? Should check to see if the two paths (minus the last item) are equal?

File: src/js/state/QueryLibrary/types.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
+export type QueryLibraryState = Group
+
+export interface Query {
+  id: string
+  name: string
+  value: string
```
Comment (jameskerr) [inline src/js/state/QueryLibrary/types.ts:6]: What do you think about naming this field `zql` instead of `value`?

**Review by jameskerr — APPROVED** [review]
Mason, this is a great foundation. I'm glad we have those thorough tests as well. I have 2 questions for ya below.

File: src/js/state/QueryLibrary/types.ts (hunk 4 above)
Comment (mason-fish) [reply src/js/state/QueryLibrary/types.ts:6]: Sure, that works

File: src/js/state/QueryLibrary/reducer.ts (hunk 3 above)
Comment (mason-fish) [reply src/js/state/QueryLibrary/reducer.ts:74]: Oh thank you, yes this is not enough as is. will fix

Comment (mason-fish) [reply src/js/state/QueryLibrary/reducer.ts:74]: actually, this does work as is: removing before or after only matters if it IS the same group. But I still prefer the handling to be only during the special case so am adding an isEqual check in the next commit

File: src/js/state/QueryLibrary/reducer.ts [hunk 5]
```diff
@@ -71,7 +71,7 @@ const moveItem = (
   // cause an off by one issue since the destination index will be affected after
   // removal (e.g. an item cannot be moved to the end of its current group because of this).
   // For this situation we instead remove the item first, and then insert its copy
-  if (srcItemPath.length === destItemPath.length) {
+  if (isEqual(initial(srcItemPath), initial(destItemPath))) {
```
Comment (jameskerr) [inline src/js/state/QueryLibrary/reducer.ts:74]: I didn't know about that "initial" function. Very cool

**Review by jameskerr — COMMENTED** [review]
I was thinking about this this week and wondered if the QLIB prefix was a bit too jargony. I think a prefix like QUERIES would be more obvious for a new user.

```
QUERIES_ADD_ITEM
QUERIES_REMOVE_ITEM
QUERIES_MOVE_ITEM
etc..
```

**Comment by mason-fish** [comment]
@jameskerr I agree qlib is a very insider term. I'm only using it in code for var names and for the action strings, and never show it to users. I shortened it to save some typing and I think it's mostly used within folders that have the name expanded so the context for future devs should be clear, but if you feel strongly I can change

**Comment by jameskerr** [comment]
Thanks for your comment Mason. You know, it appears I do feel strongly enough to request the change to be QUERIES.  I think the top-level folder in the state should also match and just be "queries".  Thanks!

**Comment by jameskerr** [comment]
Thanks for making that name change for me!

**Pull request merged**